[DRAFT][compiler] Introduce FuseGRUPass

BalyshevArtem opened the pull request:
This pr introduces FuseGRUPass for fusing gru pattern into single CircleGRU op.

for issues: https://github.com/Samsung/ONE/issues/12263 and for #13365

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

Commit:
[DRAFT][compiler] Introduce FuseGRUPass

This pr introduces FuseGRUPass for fusing gru pattern into single CircleGRU op.

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

chunseoklee commented:
@BalyshevArtem After applying this pass to gru model, while_cond / while_body subgraph are still alive in resulting graph.

BalyshevArtem commented:
> @BalyshevArtem After applying this pass to gru model, while_cond / while_body subgraph are still alive in resulting graph.

Hmmm
For me it is worked (I used your python script to obtain model and then convert it into circle and then made fusing)

![Screenshot from 2024-08-07 11-56-53](https://github.com/user-attachments/assets/6e6c8822-11d1-4d8a-b4eb-868df861731f)

BalyshevArtem commented:
@chunseoklee, I checked your models from [GRU_fusegru.zip](https://github.com/user-attachments/files/16518720/GRU_fusegru.zip).
I did not immediately understand what you mean, but now I understand, yes, this pass does not remove additional subgraphs, a new pass is needed for this task, only fusing takes place here (which works as you can see in above screen).

Commit:
[DRAFT][compiler] Introduce EliminateDeadSubgraphPass

This pr introduces EliminateDeadSubgraphPass for removing dead subgraph.

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

BalyshevArtem commented:
@chunseoklee, I have added the part that is responsible for removing dead graphs, please check :)

seanshpark commented on compiler/luci/pass/src/EliminateDeadSubgraphPass.cpp:
```diff
@@ -0,0 +1,148 @@
+{
+  bool changed = false;
+
+  if (m->size() == 0)
+    throw std::invalid_argument("No any graphs");
```
this is logically invalid case but we don't need to throw to stop the program.
just return false would be OK.

BalyshevArtem replied:
Done

chunseoklee commented on compiler/luci/pass/src/FuseGRUPass.cpp:
```diff
@@ -0,0 +1,674 @@
+  circle_gru->input(_p->_ifm);
+  circle_gru->hidden_hidden(weight_hh_cloned);
+  circle_gru->hidden_input(weight_ih_cloned);
+  circle_gru->hidden_hidden_bias(bias_hh_cloned);
```
Those two bias tensor are not well generated in the fused GRU.

BalyshevArtem replied:
> -> since original model's FC does not have bias.

Yes, that is the reason, they are empty is FC doesn't have bias

seanshpark commented:
To check pass(es) is(are) working correctly,
- add one more more tflite recipe(s) with decomposed Gelu
- add to `compiler/circle2circle-dredd-recipe-test/test.lst
- add to `compiler/luci-pass-value-py-test/test.lst`
- update necessary modules

Commit:
[recipe] Add recipe for Decomposed GRU.

This pr adds recipe for decomposed GRU.

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

Commit:
[circle2circle-dredd-recipe-test] Add fuse_gru test

This commit adds fuse_gru test to circle2circle-dredd-recipe-test.

ONE-DCO-1.0-Signed-off-by: Artem Balyshev <a.balyshev@samsung.com>

Commit:
small fixes

BalyshevArtem commented:
> To check pass(es) is(are) working correctly,
> 
>     * add one more more tflite recipe(s) with decomposed GRU
> 
>     * add to `compiler/circle2circle-dredd-recipe-test/test.lst
> 
>     * add to `compiler/luci-pass-value-py-test/test.lst`
> 
>     * update necessary modules

I added tflite decomposed GRU recipe and circle2circle-dredd-recipe-test. 
For luci-pass-value-py-test we need support this op in the luci-interpreter, right? Can we postpone the supporting of this test for later, after completing the main work on supporting this kernel in onert-micro and its training?

seanshpark commented:
> For luci-pass-value-py-test we need support this op in the luci-interpreter, right? 

Yes.

> Can we postpone the supporting of this test for later, after completing the main work on supporting this kernel in onert-micro and its training?

How can you be certain without value check that the conversion pass is working correctly?
I think it's up to you for this. Anyway if you don't add the test, no one will.

@chunseoklee , I'd like to hear your opinion.

BalyshevArtem commented:
> How can you be certain without value check that the conversion pass is working correctly?

I checked manually the result for onert-micro. After finishing with onert-micro supporting we will ad it into luci-interpreter :)

chunseoklee commented:
> @chunseoklee , I'd like to hear your opinion.

> I checked manually the result for onert-micro. After finishing with onert-micro supporting we will ad it into luci-interpreter :)

AFAIU, GRU kernel will be added into luci-interpreter after a while. Am I right ?

seanshpark commented:
>  we will ad it into luci-interpreter :)

OK, adding a issue for this task will do for now.
